### ISSUE TITLE
High CPU usage in less demanding games since #4661

### ISSUE THREAD

**Issue by wilkovatch** [RPCS3/rpcs3#4702]
Since #4661 (due to commits be13a77 and d48f391) less demanding games consume much more CPU than they used to. (a couple of examples: with the very same settings on my i5-3330 the CPU usage in Sonic CD went from 6% to 30%, in Hatsune Miku Project DIVA F 2nd it went from 40% to 60%)

FPS are not affected in any way, even in more demanding games. (for example the average FPS in Persona 5 remained the same (14-30 depending on the situation))

From my understanding (I might be wrong) this happens because of the sleep management: each sleeping thread yields instead of waiting if the remaining sleep time is less than 500 microseconds. If all threads are sleeping then the yield has no effect, and the thread keeps looping, wasting energy. (I suppose that in less demanding games this happens during the time between the completion of a frame and the beginning of the next)

Lowering the yield treshold (host_min_quantum) from 500 to 10 microseconds avoids the increased CPU consumption, but I'm not sure if it's safe. (in the pull request kd-11 mentioned possible issues such as freezing and unpredictable performance)

Performance stats in Project Diva F 2nd's main menu with host_min_quantum set to 10 and 500:
![diff](https://user-images.githubusercontent.com/29335145/40881787-8f7229d6-66d0-11e8-81ea-f452c40d21bb.jpg)

System information:
OS: Windows 10 Pro 17134.48
CPU: Intel Core i5-3330
GPU: Radeon RX 480 4GB
RAM: 12 GB

Logs of brief runs of Sonic CD with host_min_quantum set to 10 and 500 (probably useless in this case, included for completeness):
[logs.zip](https://github.com/RPCS3/rpcs3/files/2065514/logs.zip)

**Comment by kd-11** [comment]
Windows just cannot measure smaller values than 500. Its an OS problem.
Here are the sleep timings measured on windows with your hack:
```
Application started
Calculating baseline delay...(0us)
Testing usleep(0)...
    Latency = 0us
Testing usleep(1)...
    Latency = 2us
Testing usleep(10)...
    Latency = 10us
Testing usleep(50)...
    Latency = 501us
Testing usleep(100)...
    Latency = 501us
Testing usleep(200)...
    Latency = 501us
Testing usleep(300)...
    Latency = 501us
Testing usleep(400)...
    Latency = 501us
Testing usleep(500)...
    Latency = 528us
Doing the special tests...
Testing usleep(250)...
    Latency = 501us
Testing usleep(280)...
    Latency = 501us
Testing usleep(290)...
    Latency = 501us
Testing usleep(299)...
    Latency = 501us
Testing usleep(301)...
    Latency = 501us
Testing usleep(310)...
    Latency = 501us
Testing usleep(320)...
    Latency = 501us
Testing usleep(600)...
    Latency = 1002us
Application finished
```
If you try to request an 11 microsecond delay, you get kicked up to 501 microseconds. Same goes for every value between 11 and 500. Accuracy takes a bit more work, this is normal with emulators. I will lower the quantum for linux where tests show it can work at around 50-100 microseconds, but for windows its staying as is.

**Comment by kd-11** [comment]
[timer.zip](https://github.com/RPCS3/rpcs3/files/2065714/timer.zip)
Test sample used to calculate microsleep wake times for those interested. And yes, the ps3 is accurate to within 2-8 microseconds across the entire range (with 300 being the only exception, exactly 300 triggers some kind of system bug and sleeps for random duration), see comments in https://github.com/RPCS3/rpcs3/pull/4657

**Comment by Nekotekina** [comment]
I think this is valid issue.

**Comment by Zangetsu38** [comment]
I can confirm that, tested on P5 and Nier with 10, for nier no change speed, but i have stable ppu/spu usage in very low value, and for P5 i have very big stable speed Menu run 60 fps all time and school totaly stable in 28/30 fps.
in 500 for P5 it is more yoyo fps with 45 to 60 in menu and 23 to 30 in school but get 30 fps more rarely, for nier is same yoyo but with usage ppu spu, with 01 to 20 without any reason.

**Comment by wilkovatch** [comment]
Okay, I understand now. So basically it's either high accuracy at the cost of wasted CPU power, or energy saving at the cost of inaccuracy. A compromise could be adding a configuration option for this, although I understand that it would be yet another option that people could mess with, causing problems and complaints.

Anyway, I noticed that setting the treshold to 299 microseconds still avoids the CPU waste (at least on my PC), while reducing the average delay inaccuracy. (depending on the sleep time and the nearest 500us multiple) Setting it to 300 or more instead triggers the CPU waste. (at least in the two games I tested)

Performance stats in Project Diva F 2nd's main menu with host_min_quantum set to 299 and 300:
![diff2](https://user-images.githubusercontent.com/29335145/40886222-78487e7c-6734-11e8-9169-64637c1dca6f.jpg)

Timer test with host_min_quantum set to 299 (better than when using 10, albeit still inaccurate):
```
Application started
Calculating baseline delay...(0us)
Testing usleep(0)...
    Latency = 0us
Testing usleep(1)...
    Latency = 2us
Testing usleep(10)...
    Latency = 11us
Testing usleep(50)...
    Latency = 51us
Testing usleep(100)...
    Latency = 101us
Testing usleep(200)...
    Latency = 201us
Testing usleep(300)...
    Latency = 498us
Testing usleep(400)...
    Latency = 503us
Testing usleep(500)...
    Latency = 509us
Doing the special tests...
Testing usleep(250)...
    Latency = 251us
Testing usleep(280)...
    Latency = 281us
Testing usleep(290)...
    Latency = 291us
Testing usleep(299)...
    Latency = 300us
Testing usleep(301)...
    Latency = 498us
Testing usleep(310)...
    Latency = 505us
Testing usleep(320)...
    Latency = 507us
Testing usleep(600)...
    Latency = 976us
Application finished
```

**Comment by kd-11** [comment]
@Zangetsu38 small deviations upto 10 units are still correct. You can see ps3 results in the other thread, deviation of upto around 10 units is ok for large values.

Either way, Neko is considering a different approach to timers that should allow the 'normal' sleep to have better precision.

**Comment by wilkovatch** [comment]
It turns out that the culprit is the special case `sleep_time == 300`, handling it separately while keeping the treshold at 500us fixes the CPU usage problem, while keeping the sleep delays accurate. (except for the case 300 of course)

Performance stats in Sonic CD's intro screen before and after handling the case 300 (in both cases host_min_quantum is 500):
![test](https://user-images.githubusercontent.com/29335145/41034989-9b8a2a4e-698b-11e8-8bf8-4b00ffbeccdc.png)


Timer test output after handling the case 300:
```
Application started
Calculating baseline delay...(0us)
Testing usleep(0)...
    Latency = 0us
Testing usleep(1)...
    Latency = 2us
Testing usleep(10)...
    Latency = 11us
Testing usleep(50)...
    Latency = 51us
Testing usleep(100)...
    Latency = 101us
Testing usleep(200)...
    Latency = 201us
Testing usleep(300)...
    Latency = 498us
Testing usleep(400)...
    Latency = 401us
Testing usleep(500)...
    Latency = 501us
Doing the special tests...
Testing usleep(250)...
    Latency = 251us
Testing usleep(280)...
    Latency = 281us
Testing usleep(290)...
    Latency = 291us
Testing usleep(299)...
    Latency = 300us
Testing usleep(301)...
    Latency = 303us
Testing usleep(310)...
    Latency = 311us
Testing usleep(320)...
    Latency = 321us
Testing usleep(600)...
    Latency = 869us
Application finished
```

(side note: the case 600 was inaccurate even before for me, the only way to fix it is increasing the yield treshold above 500, which by the way does not seem to affect performance or CPU usage in any way)

**Comment by Rozzemak** [comment]
I heard of this 500us limitation quite some time ago...
[Interesting article](https://msdn.microsoft.com/en-us/library/windows/desktop/dn553408%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396)
So which timer is used in rpcs3 ? QPC or RDTSC ?

**Comment by wilkovatch** [comment]
Solved by #6207 (setting accuracy as host fixes the issue), closing.
